src/Events/Manager.php

```diff
@@ -177,59 +177,75 @@ public function enablePriorities(bool $enablePriorities): void
      * $eventsManager->fire("db", $connection);
      *```
      *
-     * @param string     $eventType
-     * @param object     $source
+     * @param string $eventType
+     * @param object $source
      * @param mixed|null $data
-     * @param bool       $cancelable
-     *
+     * @param bool $cancelable
      * @return mixed
-     * @throws Exception
      */
     public function fire(
         string $eventType,
         object $source,
         mixed $data = null,
-        bool $cancelable = true
+        bool $cancelable = true,
     ): mixed {
         if (empty($this->events)) {
             return null;
         }
 
-        // All valid events must have a colon separator
-        if (!str_contains($eventType, ':')) {
-            throw new Exception('Invalid event type ' . $eventType);
-        }
-
-        $eventParts = explode(':', $eventType);
-        $type       = $eventParts[0];
-        $eventName  = $eventParts[1];
-        $status     = null;
-
         // Responses must be traced?
         if (true === $this->collect) {
             $this->responses = [];
         }
 
-        // Create the event context
-        $event = new Event($eventName, $source, $data, $cancelable);
+        $status   = null;
+        $psrEvent = false;
 
-        // Check if events are grouped by type
-        $fireEvents = $this->events[$type] ?? null;
-        if (is_object($fireEvents)) {
-            // Call the events queue
-            $status = $this->fireQueue($fireEvents, $event);
+        // phalcon style events look like this: <event_group>:<event_type>
+        // PSR style events use class name as an event name by default
+        if (!($data instanceof PsrEventInterface) && str_contains($eventType, ':')) {
+            [$type, $eventName] = explode(':', $eventType);
+
+            if ($data instanceof Event) {
+                $event = $data;
+            } else {
+                $event = new Event($eventName, $source, $data, $cancelable);
+            }
+
+            // Check if events are grouped by type
+            $fireEvents = $this->events[$type] ?? null;
+            if (is_object($fireEvents)) {
+                // Call the events queue
+                $status = $this->fireQueue($fireEvents, $event);
+            }
+        } else {
+            $psrEvent = true;
+            $event = new Event($eventType, $source, $data, $cancelable);
         }
 
         // Check if there are listeners for the event type itself
         $fireEvents = $this->events[$eventType] ?? null;
         if (is_object($fireEvents)) {
             // Call the events queue
-            $status = $this->fireQueue($fireEvents, $event);
+            $status = $this->fireQueue($fireEvents, $event, $psrEvent);
         }
 
         return $status;
     }
 
+    public function dispatch(object $event, ?string $name = null, ?object $source = null)
+    {
+        if ($event instanceof Event) {
```

Comment:
I wonder if this is the place for all custom logic and I should not rely on `fire()` at all to make it simple.